rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+   This command provides options to list all Nix language versions supported by the evaluator.
+
+   <details><summary>Arguments</summary>
+
+   - (+) This is for convenience to determine which features are available
+
+   </details>
+
+1. `builtins.langVersion` returns the language version used for evaluating the given expression.
+
+   <details><summary>Arguments</summary>
+
+   - (+) Could make use of it for generating Nix expressions programmatically and annotating them with the correct version
+   - (+) `builtins.langVersion` is already part of the stable interface, this way we can make it pure
+   - (-) Requires maintaining more API surface without a clear use case
+
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - Return the latest language version instead
+
+      - (-) Doesn't help to determine what is used for evaluating the current expression
+          - (+) Might provide opportunities for forward-compatible Nix code
+          - (-) Brittle and defeats the purpose of this RFC
+      - (-) Impure, the value depends on the environment
+
+   - Don't expose `builtins.langVersion` at all
+
+      - (+) No need to have this if the Nix files are versioned
+      - (+) `builtins.langVersion` is not documented and not widely used
+      - (-) Requires a Nix language version bump to implement this RFC
+
+   </details>
+
+1. Each time the language specification (currently as embodied by the Nix language evaluator) is changed, the language version must be incremented.
+
+   <details><summary>Arguments</summary>
+
+     - (+) The principled solution: guarantees reproducibility given a fixed language version
+     - (-) Implies additional overhead in development effort:
+       - Either for Nix maintainers to accommodate that practice in the release lifecycle
+         - For example, one would have to batch language changes for a version bump to limit the number of increments
+           - (+) This would be beneficial for alternative implementations in terms of churn and effort to keep up
+       - Or implementors of alternative evaluators catching up with changes
+         - (+) Specifying the language precisely via the version actually offers alternative implementations an alternative to catching up: only support a given language version
+     - (+) This essentially nudges one to organise Nix language (specification or evaluator implementation ) development to be more independent of the rest of Nix
+       This is good, since it in turn forces stronger separation of concerns and more architectural clarity
+     - (-) Prohibits best-effort attempts at evaluating expressions with possibly incompatible evaluators
+       - (+) With the proposed level of strictness, one doesn't have to rely on best effort but can instead be explicit
+     - (-) Fixing evaluator bugs (i.e., clearly unintended behavior) after releases would technically require a version bump and therefore (theoretically) cooperation by expression authors
```

Comment:
Rust has the policy that breaking code that shouldn't have worked in the first place is not a breaking change (by definition).

Reply:
Well, that depends on whether you define bugs to be part of the interface. In lack of a specification that's not the implementation itself, and given the general aspiration around Nix to keep stuff working when not deliberately changed by users, I'm inclined to do that.

Reply:
Bugs are defined as any violations of earlier made contracts and guarantees. In practice, most of the time these are unsound APIs and undefined behavior. There might be other causes as well, like incorrect algorithms.

Reply:
Pointer equality is arguably a bug, but alas…

Reply:
> [bugs are] unsound APIs

Unsound with respect to what? What is that contract or guarantee supposed to be in our case?

> [bugs are] undefined behavior

As @sternenseemann pointed out, we don't understand the language well enough to even define undefined behavior. But even then, one would have to determine if it's fine to fill in undefined gaps without bumping the language version.

> [bugs might be] incorrect algorithms

As we have seen only recently, even an obviously inconsistent serialisation of floats is something we're stuck with forever if we *really* don't want to break existing code. I don't think the argument holds up.

What I can imagine though, therefore thanks for the intellectual resistance, is that unintended changes against previous specifications can be considered a bug. In fact, the changed floating point behavior can be said to be a bug regardless of it's more sensible. If it is not said to be a bug but an intended change, it would have required a version bump.

Reply:
I didn't say these are the definitions we should use, I just showed what Rust does as an example. And yes I am aware of https://xkcd.com/1172/, and that people will depend on software's most misbehaviors. One compromise would be to allow such kind of bug fixes within n months of the feature's introduction.

> Unsound with respect to what?

Soundness is a formal property of a type system, and only because of Rust's `unsafe` does this partially become a code problem. So yes I don't think this is really applicable to Nix.

Reply:
> One compromise would be to allow such kind of bug fixes within n months of the feature's introduction.

From experience, talking about any sort of timing in open source development really doesn't make sense. One thing I could imagine, also from experience, is something like a stability staging that takes one or two releases.

(Whether scheduling releases is meaningful or not is an independent question and does not concern this proposal. There are alternatives to cutting a release at fixed intervals.)

Reply:
Outdated as references to `builtins.langVersion` were removed.